File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int max=0,currentsum=0;
        int[] prefixarr=new int[prices.length];
        prefixarr[0]=0;
        for(int i=1;i<prices.length;i++){
            prefixarr[i]=prices[i]-prices[i-1];
        }
        for(int i=1;i<prices.length;i++){
            currentsum+=prefixarr[i];
            if(currentsum<0)
                currentsum=0;
            else if(currentsum>max)
                max=currentsum;
        }
        return max;
    }
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        res = 0
        dif = -1
        crt= prices[0]
        for i in range(1, len(prices)):
            dif = prices[i]-crt
            if dif<0:
                crt = prices[i]
            else:
                res = max(res, dif)
        
        return res